Refactor/split breeds viewmodel

carlos01sword opened the pull request:
This pull request introduces a new centralized `FavoritesState` observable object to manage favorite cat breeds, refactors related view models and views to use this shared state, and improves the handling of favorite toggling and state consistency throughout the app. The changes enhance code modularity, ensure a single source of truth for favorites, and simplify favorite-related logic in the UI.

**Centralized Favorites State Management:**

* Introduced a new `FavoritesState` class as a `@MainActor` observable object to manage the list of favorite breeds, provide methods to toggle favorites, and check favorite status. This replaces previous scattered logic and ensures synchronization across views. (`CatsApp/Presentation/ViewModels/FavoritesState.swift`)
* Updated the app entry point to instantiate and inject a single `FavoritesState` instance throughout the app using `.environmentObject`. 

**View Model Refactoring:**

* Refactored `BreedsViewModel` to depend on `FavoritesState` for favorite breed management, removing its own favorite logic. The favorite breeds list and toggling now delegate to `FavoritesState`. 
* Added new `FavoritesViewModel` and `DetailsViewModel` classes that use `FavoritesState` for all favorite-related operations, further decoupling favorite logic from breed data management. 

**View Updates and UI Improvements:**

* Refactored views (`BreedTabView`, `BreedsView`, `FavoritesView`, `DetailsView`) to accept and use the shared `FavoritesState`, ensuring consistent favorite status and updating across navigation. 
* Refactored `BreedListView` to use closures for row actions (`onSelect`, `onFavorite`, `onRowAppear`) instead of requiring a row content view, simplifying the API and making it more flexible.

These changes collectively improve state management, reduce duplication, and lay the groundwork for future enhancements related to favorites.

carlos01sword committed:
refactor: Implement shared FavoritesState, DetailsViewModel and FavoritesViewModel
- Centralized favorites logic in FavoritesState and injected as root @StateObject
- Refactored BreedsViewModel into FavoritesViewModel and DetailsViewModel
- Updeated View to use the new ViewModels
- Fixed a bug in pagination trigger

carlos01sword committed:
refactor: Simplify BreedListView and Reduce View Code Duplication

- Refactored BreedListView to internalize BreedRowView creation and expose simple callbacks for selection, favorite toggle, and row appearance.
- Eliminated AnyView usage and reduced code repetition across views.

josedesousa7 commented on CatsApp/Presentation/ViewModels/BreedsViewModel/BreedsViewModel.swift:
```diff
@@ -19,16 +19,18 @@ enum LoadPhase: Equatable {
     
     @Published var catBreeds: [CatBreed] = []
     @Published var searchText: String = ""
     @Published private(set) var phase: LoadPhase = .idle
```
First things first, I would suggest you to rename this to `state`.  Might sound a bit silly, but it's the "convention" in SwiftUI. This applies for both the property name and the object.

carlos01sword replied:
Replaced the `LoadPhase` enum and `phase` variable to `state`.

josedesousa7 commented on CatsApp/Presentation/ViewModels/FavoritesState.swift:
```diff
@@ -0,0 +1,33 @@
+import SwiftData
+
+@MainActor
+final class FavoritesState: ObservableObject{
```
```suggestion
final class FavoritesState: ObservableObject {
```

josedesousa7 commented on CatsApp/Presentation/ViewModels/FavoritesViewModel/FavoritesViewModel.swift:
```diff
@@ -0,0 +1,23 @@
+
+@MainActor
+final class FavoritesViewModel: ObservableObject {
+    let favoritesState: FavoritesState
```
I think the logic to handle the favourites can all be moved to here, instead of having a separate object `FavoritesState` and in the current structure, you make this your `environmentObject`

carlos01sword replied:
Encapsulate all favorites-related logic, including toggling favorites and loading favorite breeds into the FavoritesViewModel as suggested. and also made sure to make it my `environmentObject`.

josedesousa7 commented on CatsApp/Presentation/Views/Screens/BreedsView/BreedsView.swift:
```diff
@@ -10,64 +10,71 @@ import SwiftUI
+
     var body: some View {
         NavigationStack {
-            Group {
```
Good job splitting the body of the view

carlos01sword replied:
Appreciate 🥹

josedesousa7 commented on CatsApp/Presentation/Views/Screens/DetailsView/DetailsView.swift:
```diff
@@ -9,34 +9,36 @@ import SwiftUI
+        _viewModel = StateObject(wrappedValue: DetailsViewModel(breed: breed, favoritesState: favoritesState))
+    }
+
     var body: some View {
```
Also try to split this into smaller chunks. It's easier to read 😅

carlos01sword replied:
will do 😅

carlos01sword committed:
fix: rename loadPhase enum to state to match SwiftUI conventions

carlos01sword committed:
refactor: consolidate favorites logic
- Remove FavoritesState
- Consolidate all favorites management into FavoritesViewModel,
- Make FavoritesViewModel the sole environment object for favorites across the app.
- Updated all views and view models to use FavoritesViewModel directly

carlos01sword committed:
refactor: migrate network and data layers to use async/await

carlos01sword commented:
### Migrate Networking and Data Layer to Async/Await: 

* Refactored network and repository APIs to use Swift's `async/throws` pattern instead of returning `Result` types, resulting in more idiomatic error handling and simpler call sites.
